library/Booster/Syntax/Json/Externalise.hs

```diff
@@ -10,33 +10,48 @@ module Booster.Syntax.Json.Externalise (
 ) where
 
 import Data.Foldable ()
+import Data.List (partition)
 import Data.Text.Encoding qualified as Text
 
 import Booster.Pattern.Base (externaliseKmapUnsafe)
 import Booster.Pattern.Base qualified as Internal
 import Booster.Pattern.Util (sortOfTerm)
 import Kore.Syntax.Json.Types qualified as Syntax
 
-{- | Converts an internal pattern to a pair of term and predicate in
- external format. The predicate is 'And'ed to avoid leaking
+{- | Converts an internal pattern to a triple of term, predicate and substitution in
+ external format. The predicate and substitution are 'And'ed to avoid leaking
  Json format internals to the caller.
 -}
 externalisePattern ::
     Internal.Pattern ->
-    (Syntax.KorePattern, Maybe Syntax.KorePattern)
+    (Syntax.KorePattern, Maybe Syntax.KorePattern, Maybe Syntax.KorePattern)
 externalisePattern Internal.Pattern{term = term, constraints} =
     -- need a sort for the predicates in external format
     let sort = externaliseSort $ sortOfTerm term
+        (substitutionItems, predicateItems) = partition isSubstitutionItem constraints
+        substitution =
+            if null substitutionItems
+                then Nothing
+                else Just . multiAnd sort . map (externalisePredicate sort) $ substitutionItems
         predicate =
-            if null constraints
+            if null predicateItems
                 then Nothing
-                else Just $ multiAnd sort $ map (externalisePredicate sort) constraints
-     in (externaliseTerm term, predicate)
+                else Just . multiAnd sort . map (externalisePredicate sort) $ predicateItems
+     in (externaliseTerm term, predicate, substitution)
   where
     multiAnd :: Syntax.Sort -> [Syntax.KorePattern] -> Syntax.KorePattern
     multiAnd _ [] = error "multiAnd: empty"
     multiAnd sort ps = foldl1 (Syntax.KJAnd sort) ps
 
+    isSubstitutionItem :: Internal.Predicate -> Bool
+    isSubstitutionItem (Internal.EqualsTerm lhs _rhs) = isVariable lhs
+    isSubstitutionItem (Internal.AndPredicate lhs rhs) = isSubstitutionItem lhs && isSubstitutionItem rhs
```

Comment:
@jberthold @goodlyrottenapple I've added a recursive case here, but I'm not sure it is actually needed or even wanted. What do you think? `Pattern` holds a list of `Predicate` constraints. Could any of those `Predicate`'s be nested `AndPredicate`?

Reply:
That shouldn't happen but probably best keep it in. I'm wondering whether we need to handle cases like `X ==Int ... #Equals true`?

Reply:
I though about it, and I don't know actually. I think these are actually eliminated when we run the the Booster result through the Kore simplifier. I will need to test it.

Reply:
Agree with Sam that we should keep this (we cannot guarantee this invariant for all future times).